tree/tree.go

```diff
@@ -0,0 +1,131 @@
+package tree
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"hash/fnv"
+	"io/ioutil"
+	"sort"
+
+	"github.com/treeverse/lakefs/tree/sstable"
+)
+
+const SplitFactor = 200_000 // avarage number of entries in a part file. used to compute modulo on hash of path
+
+type TreePartType struct {
+	PartName string `json:"part_name"`
+	MaxPath  string `json:"max_path"`
+}
+type treeSliceType []TreePartType
+type TreeType struct {
+	TreeID    TreeID
+	TreeSlice []TreePartType
+}
+
+type TreesRepoType map[TreeID]TreeType
+
+var treeRepository TreesRepoType
+
+type treeScanner struct {
+	TreeType
+	currentIter sstable.EntryIterator
+	currentSst  int
+	currentKey  []byte
+}
+
+func InitTreeRepository() TreesRepoType {
+	treeRepository = make(TreesRepoType, 100)
+	return treeRepository
+}
+
+func isSplitPath(path []byte) bool {
+	fnv := fnv.New32a()
+	fnv.Write(path)
+	i := fnv.Sum32()
+	return (i % SplitFactor) == 0
+}
+
+func (trees TreesRepoType) loadTreeIfNeeded(treeID TreeID) (TreeType, error) {
+	t, exists := trees[treeID]
+	if exists {
+		return t, nil
+	}
+	fName := base64.StdEncoding.EncodeToString(treeID[:]) + ".json"
+	jsonBytes, err := ioutil.ReadFile(fName)
+	if err != nil {
+		return TreeType{}, err
+	}
+	treeSlice := make(treeSliceType, 0)
+	err = json.Unmarshal(jsonBytes, &treeSlice)
+	if err != nil {
+		return TreeType{}, err
+	}
+	t = TreeType{
+		treeID,
+		treeSlice,
+	}
+	trees[treeID] = t
+	return t, nil
+}
+
+func (trees TreesRepoType) NewScanner(tree TreeID, start string) (*treeScanner, error) {
+	t, err := trees.loadTreeIfNeeded(tree)
+	if err != nil {
+		return nil, err
+	}
+	sst := findSST(t, start)
+	if sst == nil {
+		return nil, ErrPathBiggerThanMaxPath
```

Comment:
not sure when this error value is useful, its name is hard to understand

Reply:
Scanner started with path that is above the maximum path in the tree